prov/efa/src/efa_conn.c

```diff
@@ -0,0 +1,468 @@
+	if (err) {
+		if (av->domain->info_type == EFA_INFO_RDM) {
+			/* insert_implicit_av is only true for the CQ read path
+			 * which already has the SRX lock */
+			if (insert_implicit_av)
+				ofi_genlock_lock(&av->domain->srx_lock);
+			efa_conn_rdm_deinit(av, conn);
+			if (insert_implicit_av)
+				ofi_genlock_unlock(&av->domain->srx_lock);
+		}
+		goto err_release;
+	}
+
+	insert_implicit_av ? av->used_implicit++ : av->used_explicit++;
+
+	return conn;
+
+err_release:
+	if (conn->ah)
+		efa_ah_release(av->domain, conn->ah, insert_implicit_av);
+
+	conn->ep_addr = NULL;
+	err = ofi_av_remove_addr(util_av, fi_addr);
+	if (err)
+		EFA_WARN(FI_LOG_AV, "While processing previous failure, ofi_av_remove_addr failed! err=%d\n",
+			 err);
+
+	return NULL;
+}
+
+void efa_conn_release_reverse_av(struct efa_av *av, struct efa_conn *conn,
+				 bool release_from_implicit_av)
+{
+	if (release_from_implicit_av) {
+		assert(ofi_genlock_held(&av->util_av_implicit.lock));
+		efa_av_reverse_av_remove(&av->cur_reverse_av_implicit,
+					 &av->prv_reverse_av_implicit, conn);
+	} else {
+		assert(ofi_genlock_held(&av->util_av.lock));
+		efa_av_reverse_av_remove(&av->cur_reverse_av,
+					 &av->prv_reverse_av, conn);
+	}
+}
+
+void efa_conn_release_util_av(struct efa_av *av, struct efa_conn *conn,
+			      bool release_from_implicit_av)
+{
+	struct util_av *util_av;
+	struct util_av_entry *util_av_entry;
+	struct efa_av_entry *efa_av_entry;
+	char gidstr[INET6_ADDRSTRLEN];
+	fi_addr_t fi_addr;
+	int err;
+
+	if (release_from_implicit_av) {
+		assert(ofi_genlock_held(&av->util_av_implicit.lock));
+		util_av = &av->util_av_implicit;
+		fi_addr = conn->implicit_fi_addr;
+	} else {
+		assert(ofi_genlock_held(&av->util_av.lock));
+		util_av = &av->util_av;
+		fi_addr = conn->fi_addr;
+	}
+
+	util_av_entry = ofi_bufpool_get_ibuf(util_av->av_entry_pool, fi_addr);
+	assert(util_av_entry);
+	efa_av_entry = (struct efa_av_entry *) util_av_entry->data;
+
+	err = ofi_av_remove_addr(util_av, fi_addr);
+	if (err) {
+		EFA_WARN(FI_LOG_AV, "ofi_av_remove_addr failed! err=%d\n", err);
+	}
+
+	inet_ntop(AF_INET6, conn->ep_addr->raw, gidstr, INET6_ADDRSTRLEN);
+	EFA_INFO(FI_LOG_AV, "efa_conn released! conn[%p] GID[%s] QP[%u]\n",
+		 conn, gidstr, conn->ep_addr->qpn);
+
+	conn->ep_addr = NULL;
+	memset(efa_av_entry->ep_addr, 0, EFA_EP_ADDR_LEN);
+}
+
+/**
+ * @brief release an efa conn object
+ * Caller of this function must obtain av->util_av.lock or
+ * av->util_av_implicit.lock. This function obtains the SRX lock and is called
+ * from the AV removal path.
+ *
+ * @param[in]	av	address vector
+ * @param[in]	conn	efa_conn object pointer
+ * @param[in]	release_from_implicit_av		whether to release conn
+ * from implicit AV
+ * @param[in]	grab_srx_lock		whether to get the SRX lock before
+ * destroying the peer struct
+ */
+void efa_conn_release(struct efa_av *av, struct efa_conn *conn,
+		      bool release_from_implicit_av)
```

Comment:
AFAICT, all callers to this and `efa_conn_release_ah_unsafe()` are hardcoding their `release_from_implicit_av` argument.

I know these weren't implemented for this PR, but they need to eventually be refactored into implicit/explicit-specific variants. While they're relatively small and might satisfy compiler heuristics for inlining, it's not guaranteed, meaning we can't rely on the unused branches being optimized away.